### ISSUE TITLE
How is SpecMap.createKeyBasedPlugin() supposed to work?

### ISSUE THREAD

**Issue by owenconti** [swagger-api/swagger-js#1117]
Ping @ponelat 

There's a performance issue with swagger-js when loading a large (500kb+) spec. It seems `SpecMap.createKeyBasedPlugin` attempts to apply every plugin to every path in the spec. 

As a temporary fix while developing, I added a simple depth check which helped improve loading performance greatly:

```
function* traverse(obj, path, patch) {
    depth++
    if (depth > 10) {
        return
     }
    ...
```

Nothing *seems* to break, however, I'm sure something is broken somewhere that I haven't been able to test yet. 

---

So in summary, is it intentional that every plugin is applied to all paths in the spec? Is there anything we can do to improve the performance when initializing the `SpecMap`?

**Comment by ponelat** [comment]
yes, all plugins are applied to all paths. 
I'd like to see a new specmap 2.0 ( its been talked about ), not sure if this is reason enough for it. Just note, that we can't guarantee a fully resolved spec if we limit it to a certain depth ( it'll pass, it just _might_ not be resolved ).

**Comment by owenconti** [comment]
Thanks @ponelat. I understand the issue with the `depth` check. I will need to look into it more.
